training/eval/bucketing.py

```diff
@@ -0,0 +1,231 @@
+                    max_score = score
+            else:
+                if score > max_score:
+                    if len(boundaries) == bound_num:
+                        boundaries.append(bound)
+                    else:
+                        boundaries[bound_num] = bound
+                    max_score = score
+        max_score = -np.inf
+        previous_bound = boundaries[bound_num] + 0.01
+    return boundaries, max_score
+
+
+def _get_labeler_score_optimization_version(
+    df: pd.DataFrame,
+    labeler_col: str,
+    human_prefix: str = "label_",
+    n_labels: int = 5,
+):
+    """
+    Simulates evaluation scoring but for given labels number (n_labels). Used
+    for mapping buckets optimization.
+    Args:
+        df: DataFrame where each row is a one example and each column is
+            an annotator. The values should not be normalized.
+        labeler_col: name of column which contains continous values of model
+            mapped to human labels
+        human_prefix: prefix of the columns names that store the reference
+            annotations (human labels)
+        n_labels: number of labels which were considered during mapping
+    """
+    comparison_cols = [
+        col
+        for col in df.columns
+        if (col.startswith(human_prefix) and col != labeler_col)
+    ]
+    scores = []
+    for _, row in df.iterrows():
+        labeler_score = row[labeler_col]
+        comparison_scores = row[comparison_cols].values.tolist()
+        if labeler_score > n_labels:
+            labeler_score = n_labels
+        comparison_scores = [
+            n_labels if score > n_labels else score
+            for score in comparison_scores
+        ]
+        if not np.isnan(comparison_scores).all():
+            typical_score = np.nanmedian(comparison_scores)
+            score = (
+                1
+                - ((typical_score - labeler_score) ** 2) / (n_labels - 1) ** 2
+            )
+            scores.append(score)
+        else:
+            scores.append(np.nan)
+    return np.nanmean(scores)
+
+
+def _get_tr_df_wide(dataset_df, preds_df, metric):
+    """
+    Preprocess the dataset to extract training rows, add the model's predictions
+    and convert to wide format.
+    """
+    dataset_df_copy = dataset_df.copy()
+
+    # filter training split
+    tr_df = dataset_df_copy[dataset_df_copy[f"{metric}_split"] == "train"]
+
+    # filter relevant columns and convert metric column to a list.
+    tr_df = tr_df[["ArticleID", "System", metric]]
+    tr_df[metric] = tr_df[metric].apply(ast.literal_eval)
+
+    return _to_wide(tr_df, preds_df, metric=metric)
+
+
+def _to_wide(df, preds_df, metric) -> pd.DataFrame:
+    """
+    Converts a dataframe with one row per (article, system, labeler) to
+    a dataframe with one row per (article, system) and columns for each
+    labeler's label. Also adds the model's prediction for
+    the article-system pair.
+    """
+    print(preds_df.columns)
```

Comment:
should we keep this print ? - it might be good idea to switch to loguru for logging and changing this print to `logger.debug`